src/Arcus.Security.Providers.AzureKeyVault/Authentication/CertificateBasedAuthentication.cs

```diff
@@ -0,0 +1,51 @@
+﻿using System;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading.Tasks;
+using Arcus.Security.Providers.AzureKeyVault.Authentication.Interfaces;
+using GuardNet;
+using Microsoft.Azure.KeyVault;
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+
+namespace Arcus.Security.Providers.AzureKeyVault.Authentication
+{
+    /// <summary>
+    ///     Azure Key Vault <see cref="IKeyVaultAuthentication"/> by using client ID and certificate to authenticate the <see cref="IKeyVaultClient"/>.
+    /// </summary>
+    public class CertificateBasedAuthentication : IKeyVaultAuthentication
+    {
+        private readonly string _applicationId;
+        private readonly X509Certificate2 _certificate;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CertificateBasedAuthentication"/> class.
+        /// </summary>
+        /// <param name="applicationId">The identifier of the application requesting the authentication token.</param>
+        /// <param name="certificate">The certificate that is used as credential.</param>
+        public CertificateBasedAuthentication(string applicationId, X509Certificate2 certificate)
+        {
+            Guard.NotNull(applicationId, nameof(applicationId));
+            Guard.NotNull(certificate, nameof(certificate));
+
+            _applicationId = applicationId;
+            _certificate = certificate;
+        }
+
+        /// <summary>
+        ///     Authenticates with Azure Key Vault
+        /// </summary>
+        /// <returns>A <see cref="IKeyVaultClient" /> client to use for interaction with the vault</returns>
+        public Task<IKeyVaultClient> Authenticate()
+        {
+            IKeyVaultClient client = new KeyVaultClient((async (authority, resource, scope) =>
```

Comment:
We don't tend to do inline stuff like this, please move to a dedicated method similar to our other authentication options: https://github.com/arcus-azure/arcus.security/blob/master/src/Arcus.Security.Providers.AzureKeyVault/Authentication/ServicePrincipalAuthenticator.cs#L54